Index: source/common/unicode/utypes.h
===================================================================
--- source/common/unicode/utypes.h	(revision 241384)
+++ source/common/unicode/utypes.h	(working copy)
@@ -448,7 +448,7 @@
 #define U_IO_API
 #define U_TOOLUTIL_API
 #elif defined(U_COMMON_IMPLEMENTATION)
-#define U_DATA_API     U_IMPORT
+#define U_DATA_API     U_EXPORT
 #define U_COMMON_API   U_EXPORT
 #define U_I18N_API     U_IMPORT
 #define U_LAYOUT_API   U_IMPORT
