diff --git a/source/common/unicode/platform.h b/source/common/unicode/platform.h
index 07dce8a..0b40fed 100644
--- a/source/common/unicode/platform.h
+++ b/source/common/unicode/platform.h
@@ -121,8 +121,10 @@
 #define U_PF_QNX 3700
 /** Linux is a Unix-like operating system. @internal */
 #define U_PF_LINUX 4000
+/** Native Client is pretty close to Linux. @internal */
+#define U_PF_NATIVE_CLIENT 4050
 /** Android is based on Linux. @internal */
-#define U_PF_ANDROID 4050
+#define U_PF_ANDROID 4100
 /** z/OS is the successor to OS/390 which was the successor to MVS. @internal */
 #define U_PF_OS390 9000
 /** "IBM i" is the current name of what used to be i5/OS and earlier OS/400. @internal */
@@ -140,6 +142,8 @@
 #   define U_PLATFORM U_PF_ANDROID
     /* Android wchar_t support depends on the API level. */
 #   include <android/api-level.h>
+#elif defined(__native_client__)
+#   define U_PLATFORM U_PF_NATIVE_CLIENT
 #elif defined(linux) || defined(__linux__) || defined(__linux)
 #   define U_PLATFORM U_PF_LINUX
 #elif defined(__APPLE__) && defined(__MACH__)
