diff --git a/source/common/unicode/unistr.h b/source/common/unicode/unistr.h
index c6e8b44..c98aec7 100644
--- a/source/common/unicode/unistr.h
+++ b/source/common/unicode/unistr.h
@@ -34,7 +34,6 @@
 #include "unicode/ucasemap.h"
 
 struct UConverter;          // unicode/ucnv.h
-class  StringThreadTest;
 
 #ifndef U_COMPARE_CODE_POINT_ORDER
 /* see also ustring.h and unorm.h */
@@ -3491,7 +3490,6 @@ private:
     kWritableAlias=0
   };
 
-  friend class StringThreadTest;
   friend class UnicodeStringAppendable;
 
   union StackBufferOrFields;        // forward declaration necessary before friend declaration
