diff --git a/source/common/dictbe.cpp b/source/common/dictbe.cpp
index c5aa454..88e9139 100644
--- a/source/common/dictbe.cpp
+++ b/source/common/dictbe.cpp
@@ -832,11 +832,11 @@ foundBest:
 static const int32_t KHMER_LOOKAHEAD = 3;
 
 // Will not combine a non-word with a preceding dictionary word longer than this
-static const int32_t KHMER_ROOT_COMBINE_THRESHOLD = 3;
+static const int32_t KHMER_ROOT_COMBINE_THRESHOLD = 10;
 
 // Will not combine a non-word that shares at least this much prefix with a
 // dictionary word, with a preceding word
-static const int32_t KHMER_PREFIX_COMBINE_THRESHOLD = 3;
+static const int32_t KHMER_PREFIX_COMBINE_THRESHOLD = 5;
 
 // Minimum word size
 static const int32_t KHMER_MIN_WORD = 2;
