diff --git a/source/i18n/double-conversion-utils.h b/source/i18n/double-conversion-utils.h
index 1e44fcaa..e4e5a854 100644
--- a/source/i18n/double-conversion-utils.h
+++ b/source/i18n/double-conversion-utils.h
@@ -115,6 +115,8 @@ int main(int argc, char** argv) {
 #else
 #undef DOUBLE_CONVERSION_CORRECT_DOUBLE_OPERATIONS
 #endif  // _WIN32
+#elif U_PLATFORM == U_PF_BROWSER_NATIVE_CLIENT
+#undef DOUBLE_CONVERSION_CORRECT_DOUBLE_OPERATIONS
 #else
 #error Target architecture was not detected as supported by Double-Conversion.
 #endif
