diff --git a/source/common/unicode/enumset.h b/source/common/unicode/enumset.h
index bde8c455..6d7fa72b 100644
--- a/source/common/unicode/enumset.h
+++ b/source/common/unicode/enumset.h
@@ -43,7 +43,7 @@ public:
     inline int32_t contains(T toCheck) const { return get(toCheck); }
     inline void set(T toSet, int32_t v) { fBools=(fBools&(~flag(toSet)))|(v?(flag(toSet)):0); }
     inline int32_t get(T toCheck) const { return (fBools & flag(toCheck))?1:0; }
-    inline UBool isValidEnum(T toCheck) const {  return (toCheck>=minValue&&toCheck<limitValue); }
+    inline UBool isValidEnum(T toCheck) const {  return ((uint32_t)toCheck>=minValue&&(uint32_t)toCheck<limitValue); }
     inline UBool isValidValue(int32_t v) const { return (v==0||v==1); }
     inline const EnumSet<T,minValue,limitValue>& operator=(const EnumSet<T,minValue,limitValue>& other) {
         fBools = other.fBools;
